### ISSUE TITLE
Improve error message when ingesting pcap results in empty space

### ISSUE THREAD

**Issue by siskojr** [brimdata/brim#790]
When ingesting big pcap files, space gets created but there is no data inside the space.

The request is to provide some message to the user that indicates that "there is an undefined problem in the pcap file".

The user @kav10dex reported this problem. After the analysis below pointing to errors associated to mergecap, asked user if  the problem could be combination of different formats in the packet file. The user confirmed that using the tool pcapfix <file> solved the issue and ingest in Brim was successful after that.

Error reproduced using big files (25GB, 40GB, 75GB).
There is no error message to the user. There is no message at all and ingest process finishes apparently correctly. This ingest is very fast which is not consistent with the size of the file (same situation also for the user reporting the problem).

Result is: 

1. Processing
![image.png](https://images.zenhubusercontent.com/5e8b7dd61f02b3461f7761d3/900b7c6f-44b1-47b0-a3d4-5f20ce907135)
2. Finishes with no data
![image.png](https://images.zenhubusercontent.com/5e8b7dd61f02b3461f7761d3/1e111d0a-4a3e-414d-8b95-a71a6cd50320)

I have been able to reproduce the problem using a 25 GB pcap file made of smaller ones. The smaller files successfully ingest in Brim, but not the mergecap file. The individual files are of pcap format. The merged file is pcapng and this generates this problem. If I mergecap -F pcap, the result is that the file ingests correctly.

The error that Developer tools provides is:

```
 /Applications/Brim.a…cuteUidSearch.js:26 Uncaught TypeError: Cannot read property 'correlationId' of undefined
    at /Applications/Brim.a…cuteUidSearch.js:26
    at /Applications/Brim.a…unk/lib/index.js:11
    at dispatch (/Applications/Brim.a…ux/lib/redux.js:644)
    at /Applications/Brim.a…viewLogDetail.js:20
    at Object.dispatch (/Applications/Brim.a…unk/lib/index.js:11)
    at /Applications/Brim.a…ResultsTable.js:120
(anonymous)	@	/Applications/Brim.a…cuteUidSearch.js:26
(anonymous)	@	/Applications/Brim.a…unk/lib/index.js:11
dispatch	@	/Applications/Brim.a…ux/lib/redux.js:644
(anonymous)	@	/Applications/Brim.a…viewLogDetail.js:20
(anonymous)	@	/Applications/Brim.a…unk/lib/index.js:11
(anonymous)	@	/Applications/Brim.a…ResultsTable.js:120
setTimeout (async)		
(anonymous)	@	/Applications/Brim.a…bouncedEffect.js:14
commitHookEffectList	@	/Applications/Brim.a…evelopment.js:22030
commitPassiveHookEffects	@	/Applications/Brim.a…evelopment.js:22064
callCallback	@	/Applications/Brim.a….development.js:336
invokeGuardedCallbackDev	@	/Applications/Brim.a….development.js:385
invokeGuardedCallback	@	/Applications/Brim.a….development.js:440
flushPassiveEffectsImpl	@	/Applications/Brim.a…evelopment.js:25392
unstable_runWithPriority	@	/Applications/Brim.a….development.js:697
runWithPriority$2	@	/Applications/Brim.a…evelopment.js:12149
flushPassiveEffects	@	/Applications/Brim.a…evelopment.js:25361
performSyncWorkOnRoot	@	/Applications/Brim.a…evelopment.js:24251
(anonymous)	@	/Applications/Brim.a…evelopment.js:12199
unstable_runWithPriority	@	/Applications/Brim.a….development.js:697
runWithPriority$2	@	/Applications/Brim.a…evelopment.js:12149
flushSyncCallbackQueueImpl	@	/Applications/Brim.a…evelopment.js:12194
flushSyncCallbackQueue	@	/Applications/Brim.a…evelopment.js:12182
batchedUpdates$1	@	/Applications/Brim.a…evelopment.js:24392
notify	@	/Applications/Brim.a…/Subscription.js:29
notifyNestedSubs	@	/Applications/Brim.a…/Subscription.js:71
handleChangeWrapper	@	/Applications/Brim.a…/Subscription.js:76
dispatch	@	/Applications/Brim.a…ux/lib/redux.js:228
(anonymous)	@	/Applications/Brim.a…unk/lib/index.js:14
(anonymous)	@	/Applications/Brim.a…rs/initSearch.js:87
emit	@	events.js:208
onMessage	@	init.ts:50

```

I have also verified that the problem is in the "zeek" process that finishes very quickly but does not generate the typical log files for the associated pcap file, only packet_filter.log

```
Franciscos-MacBook-Pro:kav_bug javierrodriguez$ time zeek -C -r ../kav25.pcap 

real	0m0.855s
user	0m0.437s
sys	0m0.192s
```

Zeek only generates packet_filter.log

```
Franciscos-MacBook-Pro:logs javierrodriguez$ more packet_filter.log 
#separator \x09
#set_separator  ,
#empty_field    (empty)
#unset_field    -
#path   packet_filter
#open   2020-05-17-18-33-16
#fields ts      node    filter  init    success
#types  time    string  string  bool    bool
1589765596.662601       zeek    ip or not ip    T       T
#close  2020-05-17-18-33-16

```

The request is to report back to the user a message indicating, as an example, that the pcap file has some problem.

Ref: https://sourceforge.net/projects/pcapfix/ repairs corrupted packet files.

**Comment by philrz** [comment]
@siskojr confirmed offline the three publicly-available pcaps with which he's been able to repro this issue:

http://mawi.wide.ad.jp/mawi/samplepoint-F/2015/201508051400.dump.gz
https://download.netresec.com/pcap/maccdc-2011/maccdc2011_00010_20110312194033.pcap.gz
https://download.netresec.com/pcap/maccdc-2011/maccdc2011_00013_20110312202724.pcap.gz

I've downloaded and uncompressed these. Then I `mergecap` them to both versions:

```
$ mergecap -w brim-790.pcapng 201508051400.dump maccdc2011_00010_20110312194033.pcap maccdc2011_00013_20110312202724.pcap
$ mergecap -F pcap -w brim-790.pcap 201508051400.dump maccdc2011_00010_20110312194033.pcap maccdc2011_00013_20110312202724.pcap
```

I first reproduced the issue using the Zeek that's embedded with Brim `v0.9.1`, but I've also now reproduced it with "stock" Zeek v3.1.3 compiled on my Mac. Like reported above by @siskojr, with the pcapng version, Zeek quietly exits almost immediately with no error messages, a "success" exit code of 0, and just a few log files with no useful debug info in them.

```
$ /usr/local/zeek-3.1.3/bin/zeek -C -r ../brim-790.pcapng local
WARNING: No Site::local_nets have been defined.  It's usually a good idea to define your local networks.

$ echo $?
0

$ ls -l
total 80
-rw-r--r--  1 phil  staff  30422 May 18 11:12 loaded_scripts.log
-rw-r--r--  1 phil  staff    254 May 18 11:12 packet_filter.log
-rw-r--r--  1 phil  staff    694 May 18 11:12 stats.log
```

With the regular pcap version, it creates the usual huge set of Zeek logs.

Given that pcapfix has been shown to address these issues, and hence corruption is suspected, I decided to try other packet-reading tools to see if anything else choked on it. `tshark` does not complain, but `tcpdump` does. Maybe it's a `libpcap` problem, but for some reason Zeek is not presenting the error the same way `tcpdump` is?

```
$ tcpdump -r ../brim-790.pcapng 
reading from PCAP-NG file ../brim-790.pcapng
tcpdump: pcap_loop: invalid packet capture length 260, bigger than snaplen of 96

$ tcpdump --version
tcpdump version tcpdump version 4.9.3 -- Apple version 90.100.1
libpcap version 1.9.1
LibreSSL 2.8.3
```

I'll open a Zeek issue and see what they make of it.

**Comment by philrz** [comment]
While I've started to deal with a tactical problem by opening to the linked Zeek issue (and may do the same with a `mergecap` issue), _this_ issue lives on to address the original suggestion that we pop up some kind of informative message when any pcap import results in zero Zeek logs. Exact wording is TBD, but @siskojr's suggestions seem reasonable: Hinting that there may be a problem with the pcap and suggesting that the user investigate its health with other tools (& probably also suggesting they open an issue with us whether or not they figure it out).

**Comment by philrz** [comment]
I've also opened https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=16573 regarding mergecap having generated a pcapng that was rejected by tcpdump as shown above.

Separately, @siskojr asked the user who originally reported the problem to try reading their pcapng with tcpdump, since while the Zeek-centric symptom of our repro matched their symptom, we don't have the user's pcapng and hence can't say with certainty they have the same root cause. Indeed, the word back is that the user's symptom is slightly different:

> `tcpdump: pcap_loop: An interface has a snapshot length xxxx different from the type of the first interface.`

For this one, web searches turn up a fixed libpcap issue https://github.com/the-tcpdump-group/libpcap/issues/832. As our user is using Windows (as evidenced by some Slack conversations), I believe they'll see a fix for their specific issue in Brim soon. That libpcap issue has been fixed (I've confirmed with my tcpdump v4.9.3 that uses libpcap v1.9.1 that the symptom does not repro with the [full.pcapng.tar.gz](https://github.com/the-tcpdump-group/libpcap/files/3458691/full.pcapng.tar.gz) that's attached to that issue) but the Windows releases of Brim up until now have had to depend on old WinPcap versions that pre-date that fix. The next Brim release for Windows is expected to include a current libpcap (tracked by https://github.com/brimsec/brim/issues/563). I've confirmed with an early build of that Zeek artifact for Windows that the [full.pcapng.tar.gz](https://github.com/the-tcpdump-group/libpcap/files/3458691/full.pcapng.tar.gz) can be processed whereas our prior artifact rejected it, so we'll confirm with the user when the new Windows release is GA.

Revisiting the previous theme, _this_ issue still lives on to address the original suggestion that we pop up some kind of informative message when any attempted pcap import results in zero Zeek logs. Even if we manage to whack-a-mole these cases as we find them, the fact we've stumbled onto two root causes indicates we've probably not seen the last of them, so guiding the user for checking for pcap corruption does seem responsible.

**Comment by philrz** [comment]
I think this one can be closed as no longer relevant.

Just to revisit the state of the original issue, the combination of the Zeek fix linked to above and better delivery of error messages in Brim shows the following error is now correctly shown to the user. As of Brim commit 1ee21a8:

![image](https://user-images.githubusercontent.com/5934157/184696881-bc1c4fd7-06c5-41dd-8de2-bfdac6ba5bcc.png)

Then there's the subject of whether the app should have specific messages if a pcap doesn't produce Zeek logs. Since the time this issue was opened, the app is more loosely coupled with Zeek. While Zeek is still bundled and invoked by default on imported pcaps, in the [Brim wiki](https://github.com/brimdata/brim/wiki) we now explain how this bundling and processing is done by the separate [Brimcap](https://github.com/brimdata/brimcap) utility that's shipped with Brim. There's also a separate [Brimcap wiki](https://github.com/brimdata/brimcap/wiki) that includes [Troubleshooting](https://github.com/brimdata/brimcap/wiki/Troubleshooting) tips as well as a [Customization](https://github.com/brimdata/brimcap/wiki/Custom-Brimcap-Config) article that provides more detail on how Brimcap invokes Zeek and other tools. In conclusion, if an issue like this original one were to surface, we expect users would better be able to leverage the docs to determine it's a problem with Zeek itself.

Closing.
